File: us_market/output/sniper/porto_builder.js
// ── Proxy state ──
const _state = { selected: {} };
const state = new Proxy(_state, {
  set(target, prop, value) {
    target[prop] = value;
    if (prop === 'selected') recalculateAllocation();
    return true;
  }
});

function toggleStock(ticker) {
  const newSel = { ...state.selected };
  if (newSel[ticker]) {
    delete newSel[ticker];
  } else {
    const s = STOCKS.find(s => s.ticker === ticker);
    if (s) newSel[ticker] = { score: s.score, tf: s.tf };
  }
  state.selected = newSel;
  updateCardUI(ticker);
}

function recalculateAllocation() {
  const sel   = state.selected;
  const ticks = Object.keys(sel);
  const total = ticks.reduce((sum, t) => sum + sel[t].score, 0);
  ticks.forEach(ticker => {
    const alloc = total > 0 ? (sel[ticker].score / total) * 100 : 0;
    sel[ticker].alloc = alloc;
    const card = document.querySelector('[data-ticker="' + ticker + '"]');
    if (card) {
      const el = card.querySelector('.alloc-val');
      if (el) el.textContent = 'Alloc: ' + alloc.toFixed(1) + '%';
    }
  });
  updateFooter();
}

function updateCardUI(ticker) {
  const card  = document.querySelector('[data-ticker="' + ticker + '"]');
  if (!card) return;
  const isSel = !!state.selected[ticker];
  card.classList.toggle('selected', isSel);
  const chk = card.querySelector('.check-icon');
  if (chk) chk.textContent = isSel ? '\u2713' : '';
  const allocEl = card.querySelector('.alloc-val');
  if (allocEl && !isSel) allocEl.textContent = 'Alloc: \u2014';
}

function updateFooter() {
  const sel   = state.selected;
  const ticks = Object.keys(sel);
  const count = ticks.length;
  document.getElementById('footerCount').textContent = count;
  const chipsEl = document.getElementById('footerChips');
  const evEl    = document.getElementById('footerEv');
  if (count === 0) {
    chipsEl.innerHTML = '<span class="footer-empty">No stocks selected</span>';
    evEl.textContent  = '\u2014';
    evEl.style.color  = 'var(--dim)';
    return;
  }
  let totalEv = 0;
  ticks.forEach(ticker => {
    const s = STOCKS.find(s => s.ticker === ticker);
    if (s) totalEv += s.ev[sel[ticker].tf || '20D'];
  });
  const avgEv = totalEv / count;
  evEl.textContent = (avgEv >= 0 ? '+' : '') + avgEv.toFixed(2) + '%';
  evEl.style.color  = avgEv >= 0 ? 'var(--green)' : 'var(--red)';
  chipsEl.innerHTML = ticks.map(t => {
    const alloc = sel[t].alloc || 0;
    return '<span class="footer-chip">' + t + ' ' + alloc.toFixed(1) + '%</span>';
  }).join('');
}

// ── TF change ──
function changeTf(ticker, tf) {
  const s = STOCKS.find(s => s.ticker === ticker);
  if (!s) return;
  s.tf = tf;
  const card = document.querySelector('[data-ticker="' + ticker + '"]');
  if (card) {
    const evEl = card.querySelector('.ev-val');
    if (evEl) {
      const ev = s.ev[tf];
      evEl.textContent = 'EV: ' + (ev >= 0 ? '+' : '') + ev.toFixed(2) + '%';
    }
  }
  if (state.selected[ticker]) {
    const newSel = { ...state.selected };
    newSel[ticker].tf = tf;
    state.selected = newSel;
  }
}

// ── Stress test ──
let _stressTicker = null;
let _stressTf     = '20D';

function openStress(ticker) {
  _stressTicker = ticker;
  const s = STOCKS.find(s => s.ticker === ticker);
  _stressTf = s ? s.tf : '20D';
  renderStress();
  document.getElementById('stressOverlay').classList.add('active');
}

function switchStressTf(tf) {
  _stressTf = tf;
  renderStress();
}

function renderStress() {
  const s = STOCKS.find(s => s.ticker === _stressTicker);
  if (!s) return;
  document.getElementById('stressTicker').textContent = s.ticker + ' \u2014 ' + s.company;
  const tabs = ['5D', '10D', '20D'].map(tf =>
    '<button class="stress-tf-btn' + (tf === _stressTf ? ' active' : '') + '" ' +
    'onclick="switchStressTf(\'' + tf + '\')">' + tf + '</button>'
  ).join('');
  document.getElementById('stressTabs').innerHTML = tabs;
  const tf  = _stressTf;
  const wr  = s.wr[tf];
  const lr  = 100 - wr;
  const aw  = s.aw[tf];
  const al  = s.al[tf];
  const ev  = s.ev[tf];
  const n   = s.sample;
  function col(v)   { return v >= 0 ? 'var(--green)' : 'var(--red)'; }
  function wrCol(v) {
    if (v >= 60) return 'var(--green)';
    if (v >= 50) return 'var(--green2)';
    if (v >= 40) return 'var(--orange)';
    return 'var(--red)';
  }
  document.getElementById('stressGrid').innerHTML =
    '<div class="stress-cell"><div class="stress-cell-label">Win Rate</div>' +
      '<div class="stress-cell-val" style="color:' + wrCol(wr) + '">' + wr.toFixed(1) + '%</div></div>' +
    '<div class="stress-cell"><div class="stress-cell-label">Loss Rate</div>' +
      '<div class="stress-cell-val" style="color:var(--dim)">' + lr.toFixed(1) + '%</div></div>' +
    '<div class="stress-cell"><div class="stress-cell-label">Avg Win</div>' +
      '<div class="stress-cell-val" style="color:' + col(aw) + '">' + (aw >= 0 ? '+' : '') + aw.toFixed(2) + '%</div></div>' +
    '<div class="stress-cell"><div class="stress-cell-label">Avg Loss</div>' +
      '<div class="stress-cell-val" style="color:' + col(al) + '">' + (al >= 0 ? '+' : '') + al.toFixed(2) + '%</div></div>' +
    '<div class="stress-cell"><div class="stress-cell-label">EV (' + tf + ')</div>' +
      '<div class="stress-cell-val" style="color:' + col(ev) + '">' + (ev >= 0 ? '+' : '') + ev.toFixed(2) + '%</div></div>' +
    '<div class="stress-cell"><div class="stress-cell-label">Sample (N)</div>' +
    '<div class="stress-cell-val" style="color:var(--white)">' + n + '</div></div>';
}

function closeStress() {
  document.getElementById('stressOverlay').classList.remove('active');
}

// ── How overlay ──
function openHow()  { document.getElementById('howOverlay').classList.add('active'); }
function closeHow() { document.getElementById('howOverlay').classList.remove('active'); }

// ── Keyboard close ──
document.addEventListener('keydown', e => {
  if (e.key === 'Escape') { closeHow(); closeStress(); }
});

// ── Render cards ──
function renderCards() {
  const grid = document.getElementById('cardGrid');
  grid.innerHTML = STOCKS.map(s => {
    const ev     = s.ev[s.tf];
    const evSign = ev >= 0 ? '+' : '';
    const evDisp = 'EV: ' + evSign + ev.toFixed(2) + '%';
    // Split onto separate lines
    const tierLine = s.tier + '<br>' + s.sharia;
    return [
      '<div class="card" data-ticker="', s.ticker, '">',
        '<div class="zone-a zone-a-btn" data-ticker="', s.ticker, '">',
          '<div class="zone-a-inner"><span class="check-icon"></span></div>',
        '</div>',
        '<div class="zone-b">',
          '<div class="ticker-block">',
            '<a class="ticker-link" href="html/', s.ticker, '.html">', s.ticker, '</a>',
            '<div class="tier-line">', tierLine, '</div>',
          '</div>',
          '<div class="mid-block">',
            '<div class="tf-row">TF:&nbsp;',
              '<select class="tf-select" data-ticker="', s.ticker, '">',
                '<option value="5D">5 D</option>',
                '<option value="10D">10 D</option>',
                '<option value="20D" selected>20 D</option>',
              '</select>',
            '</div>',
            '<div class="score-row">Score: <span class="score-val">', s.score.toFixed(1), '</span></div>',
          '</div>',
        '</div>',
        '<div class="zone-c zone-c-btn" data-ticker="', s.ticker, '">',
          '<div class="ev-label ev-val">', evDisp, '</div>',
          '<div class="alloc-label alloc-val">Alloc: \u2014</div>',
        '</div>',
      '</div>'
    ].join('');
  }).join('');

  // Event delegation
  grid.addEventListener('click', function(e) {
    const zoneA = e.target.closest('.zone-a-btn');
    if (zoneA) { toggleStock(zoneA.dataset.ticker); return; }
    const zoneC = e.target.closest('.zone-c-btn');
    if (zoneC) { openStress(zoneC.dataset.ticker); return; }
  });

  grid.addEventListener('change', function(e) {
    if (e.target.classList.contains('tf-select')) {
      changeTf(e.target.dataset.ticker, e.target.value);
    }
  });
}

renderCards();

File: us_market/dashboard/porto_builder.py
import json
import math
import pandas as pd
from pathlib import Path
from datetime import datetime
import pytz

_CHARACTERISTIC_TIER = {
    "COMPOUNDER": "COMPDR",
    "BURST":      "BURST",
    "STEADY":     "STEADY",
    "ERRATIC":    "ERATIC",
}
_SHARIA_MAP = {
    "Halal": "Halal",
    "Not Halal": "NoHalal",
    "Doubtful": "Doubt",
    "Not Covered": "Uncover",
}

def _tier_from_characteristic(val: str) -> str:
    val_upper = str(val).upper()
    for key, tier in _CHARACTERISTIC_TIER.items():
        if key in val_upper:
            return tier
    return "ERRATIC"

def _sharia_compliance(val: str) -> str:
    if not val:  # handle None / "" / missing
        return "-"
    return _SHARIA_MAP.get(val, "-")

def _safe_float(val, fallback: float = 0.0) -> float:
    """Convert to float safely; returns fallback for NaN/Inf/empty."""
    try:
        result = float(val)
        if math.isnan(result) or math.isinf(result):
            return fallback
        return result
    except (ValueError, TypeError):
        return fallback

def _safe_str(val: object, fallback: str = "") -> str:
    """Convert to string, strip None."""
    if val is None or (isinstance(val, float) and math.isnan(val)):
        return fallback
    return str(val)

def _js_str(val: str) -> str:
    """Escape a Python string for safe embedding inside a JS double-quoted string."""
    return (
        val
        .replace("\\", "\\\\")
        .replace('"',  '\\"')
        .replace("'",  "\\'")
        .replace("\n", "\\n")
        .replace("\r", "")
    )


def build_stocks_data(summary_df: pd.DataFrame) -> list[dict]:
    """Build the STOCKS list from DataFrame — pure data, no HTML."""
    df = summary_df.sort_values("Sniper_Score", ascending=False).reset_index(drop=True)
    stocks = []

    for _, row in df.iterrows():
        ticker  = _safe_str(row["Ticker"], "UNKNOWN")
        company = _safe_str(row.get("Company", ticker))
        score   = _safe_float(row["Sniper_Score"])
        sample  = int(_safe_float(row["Sample"]))
        char    = _safe_str(row.get("Characteristic", ""))
        halal  = _safe_str(row.get("Sharia", "\u2014"), "\u2014")
        sharia   = _sharia_compliance(halal)
        tier    = _tier_from_characteristic(char)

        wr5  = _safe_float(row["WR_5"])  / 100
        wr10 = _safe_float(row["WR_10"]) / 100
        wr20 = _safe_float(row["WR_20"]) / 100

        aw5  = _safe_float(row["AvgWin_5"])
        aw10 = _safe_float(row["AvgWin_10"])
        aw20 = _safe_float(row["AvgWin_20"])
        al5  = _safe_float(row["AvgLoss_5"])
        al10 = _safe_float(row["AvgLoss_10"])
        al20 = _safe_float(row["AvgLoss_20"])

        avg5  = _safe_float(row["AVG_5"])
        avg10 = _safe_float(row["AVG_10"])
        avg20 = _safe_float(row["AVG_20"])

        ev5  = round(wr5  * aw5  + (1 - wr5)  * al5,  2)
        ev10 = round(wr10 * aw10 + (1 - wr10) * al10, 2)
        ev20 = round(wr20 * aw20 + (1 - wr20) * al20, 2)

        stocks.append({
            "id":      ticker,
            "ticker":  ticker,
            "company": company,
            "score":   round(score, 2),
            "sample":  sample,
            "tier":    tier,
            "char":    char,
            "sharia":  sharia,
            "tf":      "20D",
            "ev":  {"5D": ev5,  "10D": ev10,  "20D": ev20},
            "avg": {"5D": avg5, "10D": avg10, "20D": avg20},
            "wr":  {"5D": round(wr5  * 100, 2), "10D": round(wr10 * 100, 2), "20D": round(wr20 * 100, 2)},
            "aw":  {"5D": aw5,  "10D": aw10,  "20D": aw20},
            "al":  {"5D": al5,  "10D": al10,  "20D": al20},
            "selected": False,
            "alloc": 0,
        })

    return stocks


def build_porto_dashboard(summary_df: pd.DataFrame, timestamp: str) -> str:
    """
    Build the porto HTML page.
    CSS and JS are loaded from external files — Python only injects:
      1. The timestamp
      2. The STOCKS data array (one <script> block)
    """
    stocks = build_stocks_data(summary_df)

    # Use json.dumps for safe, valid JS — no manual escaping needed
    stocks_js = json.dumps(stocks, ensure_ascii=False)

    html = (
        '<!DOCTYPE html>\n'
        '<html lang="en">\n'
        '<head>\n'
        '<meta charset="UTF-8">\n'
        '<meta name="viewport" content="width=device-width, initial-scale=1.0">\n'
        '<title>Portfolio Strategy Builder</title>\n'
        '<link rel="icon" type="image/png" href="../../../assets/logo.png">\n'
        '<link rel="stylesheet" href="porto_builder.css">\n'
        '</head>\n'
        '<body>\n'

        # NAV — stacked group 
        '<div class="nav-group">\n'
        '<a href="../sniper/index.html" class="nav-btn nav-sniper">&#9664; Sniper</a>\n'
        '<a href="../../us_hub.html" class="nav-btn">&#9664; Hub</a>\n'
        '</div>\n'

        # HEADER
        '<div class="header">\n'
        '<h1>Portfolio<br>Strategy Builder</h1>\n'
        '<p class="ts">Generated: ' + timestamp + '</p>\n'
        '<p class="ts" style="color:#ff8c00">Todays runner ranked by sniper score</p>\n'
        '</div>\n'

        # ACTION BAR
        '<div class="action-bar">\n'
        '<span class="action-hint">&#8595; Click checkbox to select stocks</span>\n'
        '<button class="how-btn" onclick="openHow()">[?] Info</button>\n'
        '</div>\n'

        # CARD GRID (populated by JS)
        '<div class="card-grid" id="cardGrid"></div>\n'

        # HOW TO USE OVERLAY
        '<div class="overlay" id="howOverlay">\n'
        '<div class="overlay-box">\n'
        '<button class="overlay-close" onclick="closeHow()">&#10005; Close</button>\n'
        '<div class="overlay-title">How to Use</div>\n'
        '<div class="how-item"><span class="how-icon">&#9635;</span>'
        '<span><b style="color:#fff">Checkbox (orange/green block, left)</b> \u2014 Click to select or deselect a stock. '
        'Allocation % recalculates automatically across all selected stocks in real-time.</span></div>\n'
        '<div class="how-item"><span class="how-icon">&#11041;</span>'
        '<span><b style="color:#fff">Ticker Name</b> \u2014 Click to open the full Sniper Detail dashboard '
        'for that ticker: backtest history, signal table, edge &amp; quality breakdown.</span></div>\n'
        '<div class="how-item"><span class="how-icon">&#9636;</span>'
        '<span><b style="color:#fff">TF Dropdown</b> \u2014 Switch time-frame per card individually (5D / 10D / 20D). '
        'EV updates immediately on the card and in the footer basket.</span></div>\n'
        '<div class="how-item"><span class="how-icon">&#9641;</span>'
        '<span><b style="color:#fff">EV / Alloc box (right block)</b> \u2014 Click to open the Stress Test popup. '
        'Shows Win Rate, Loss Rate, Avg Win, Avg Loss, EV and Avg Return for each time-frame.</span></div>\n'
        '<div class="how-item"><span class="how-icon">&#9644;</span>'
        '<span><b style="color:#fff">Sticky Footer</b> \u2014 Displays Average EV of your basket, '
        'count of selected stocks, and a scrollable chip list showing each ticker with its '
        'score-weighted allocation %. All values update live.</span></div>\n'
        '</div></div>\n'

        # STRESS TEST OVERLAY
        '<div class="stress-overlay" id="stressOverlay">\n'
        '<div class="stress-box">\n'
        '<button class="stress-close" onclick="closeStress()">&#10005; Close</button>\n'
        '<div class="stress-title">Stress Test</div>\n'
        '<div class="stress-ticker-name" id="stressTicker">\u2014</div>\n'
        '<div class="stress-tf-tabs" id="stressTabs"></div>\n'
        '<div class="stress-grid" id="stressGrid"></div>\n'
        '<div class="stress-formula">'
        'EV : <span>(WR% &times; AvgWin) + (LR% &times; AvgLoss)</span><br>'
        'Where <span>LR% = 1 &minus; WR%</span> &nbsp;|&nbsp; <span>AvgLoss</span> is typically negative<br>'
        'Allocation : <span>Score / Total Score</span>'
        '</div>\n'
        '</div></div>\n'

        # STICKY FOOTER — row 1: avg ev + selected count | row 2: chip list
        '<div class="sticky-footer">\n'
        '<div class="footer-top">\n'
        '<div class="footer-ev">AVG EV:&nbsp;<span class="ev-val" id="footerEv" style="color:var(--dim)">\u2014</span></div>\n'
        '<div class="footer-count">Selected:<span class="cnt-val" id="footerCount">0</span></div>\n'
        '</div>\n'
        '<div class="footer-chips" id="footerChips"><span class="footer-empty">No stocks selected</span></div>\n'
        '</div>\n'

        # DATA INJECTION — only this <script> block is generated by Python
        '<script>const STOCKS = ' + stocks_js + ';</script>\n'

        # LOGIC — loaded from static external file
        '<script src="porto_builder.js"></script>\n'

        '</body></html>\n'
    )

    return html


# ── CLI runner ──────────────────────────────────────────────
if __name__ == "__main__":
    edt       = pytz.timezone("US/Eastern")
    timestamp = datetime.now(edt).strftime("%d-%m-%Y %H:%M")

    csv_path  = Path(__file__).resolve().parent / "sniper_summary.csv"
    df        = pd.read_csv(csv_path)
    html      = build_porto_dashboard(df, timestamp)

    out_path  = Path(__file__).resolve().parent / "porto_builder.html"
    with open(out_path, "w", encoding="utf-8") as f:
        f.write(html)
    print(f"✅ porto_builder.html saved → {out_path}")